vpr/src/pack/noc_aware_cluster_util.cpp

```diff
@@ -0,0 +1,91 @@
+
+#include "noc_aware_cluster_util.h"
+#include "globals.h"
+
+#include <queue>
+
+std::vector<AtomBlockId> find_noc_router_atoms() {
+    const auto& atom_ctx = g_vpr_ctx.atom();
+
+    // NoC router atoms are expected to have a specific blif model
+    const std::string noc_router_blif_model_name = "noc_router_adapter_block";
+
+    // stores found NoC router atoms
+    std::vector<AtomBlockId> noc_router_atoms;
+
+    // iterate over all atoms and find those whose blif model matches
+    for (auto atom_id : atom_ctx.nlist.blocks()) {
+        const t_model* model = atom_ctx.nlist.block_model(atom_id);
+        if (noc_router_blif_model_name == model->name) {
+            noc_router_atoms.push_back(atom_id);
+        }
+    }
+
+    return noc_router_atoms;
+}
+
+void update_noc_reachability_partitions(const std::vector<AtomBlockId>& noc_atoms) {
+    const auto& atom_ctx = g_vpr_ctx.atom();
+    auto& cl_helper_ctx = g_vpr_ctx.mutable_cl_helper();
+    const auto& high_fanout_thresholds = g_vpr_ctx.cl_helper().high_fanout_thresholds;
+    const auto& grid = g_vpr_ctx.device().grid;
+
+    t_logical_block_type_ptr logic_block_type = infer_logic_block_type(grid);
+    const char* logical_block_name = logic_block_type != nullptr ? logic_block_type->name : "";
+    const size_t high_fanout_threshold = high_fanout_thresholds.get_threshold(logical_block_name);
+
+    // get the total number of atoms
+    const size_t n_atoms = atom_ctx.nlist.blocks().size();
+
+    vtr::vector<AtomBlockId, bool> atom_visited(n_atoms, false);
+
+    cl_helper_ctx.atom_noc_grp_id.resize(n_atoms, NocGroupId::INVALID());
+
+    int noc_grp_id_cnt = 0;
+
+    for (auto noc_atom_id : noc_atoms) {
```

Comment:
Comment what you are doing here.  A breadth first search reachability analysis starting at each router, that follows input and output nets below the high_fanout_threshold, and marks each netlist block with a count indicating which cluster (NoC group) it was reached from. Atoms are only labelled once, by the first breadth first search that finds them.